### ISSUE TITLE
Support explicit WebSocket session close 

### ISSUE THREAD

**Issue by e5l** [ktorio/ktor#1946]
**Subsystem**
Client, and server possibly

**Is your feature request related to a problem? Please describe.**

```kotlin
    client.ws(host = "echo.websocket.org") {
        send(Frame.Text("Hello World"))
        close()

        incoming.consumeEach {
            println(it)
        }
    }
```

Fails with:
```
Exception in thread "main" java.util.concurrent.CancellationException: ArrayChannel was cancelled
	at kotlinx.coroutines.channels.AbstractChannel.cancel(AbstractChannel.kt:632)
	at kotlinx.coroutines.channels.ReceiveChannel$DefaultImpls.cancel$default(Channel.kt:311)
	at io.ktor.http.cio.websocket.DefaultWebSocketSessionImpl$runOutgoingProcessor$1.invokeSuspend(DefaultWebSocketSessionImpl.kt:163)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:175)
	at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined(DispatchedTask.kt:137)
	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:108)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:306)
	at kotlinx.coroutines.CancellableContinuationImpl.completeResume(CancellableContinuationImpl.kt:393)
	at kotlinx.coroutines.channels.AbstractChannel$ReceiveHasNext.completeResumeReceive(AbstractChannel.kt:918)
	at kotlinx.coroutines.channels.ArrayChannel.offerInternal(ArrayChannel.kt:83)
	at kotlinx.coroutines.channels.AbstractSendChannel.send(AbstractChannel.kt:149)
	at io.ktor.http.cio.websocket.WebSocketSession$DefaultImpls.send(WebSocketSessionJvm.kt:45)
	at io.ktor.http.cio.websocket.DefaultWebSocketSession$DefaultImpls.send(DefaultWebSocketSession.kt)
	at io.ktor.http.cio.websocket.DefaultWebSocketSessionImpl.send(DefaultWebSocketSessionImpl.kt:27)
	at io.ktor.client.features.websocket.DefaultClientWebSocketSession.send(ClientSessions.kt)
	at io.ktor.http.cio.websocket.WebSocketSessionKt.close(WebSocketSession.kt:74)
	at io.ktor.http.cio.websocket.WebSocketSessionKt.close$default(WebSocketSession.kt:72)
	at WebSocketKt$Websockets$1$1.invokeSuspend(WebSocket.kt:15)
	at WebSocketKt$Websockets$1$1.invoke(WebSocket.kt)
	at io.ktor.client.features.websocket.BuildersKt.webSocket(builders.kt:54)
	at io.ktor.client.features.websocket.BuildersKt$webSocket$1.invokeSuspend(builders.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
	at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
	at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
	at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
	at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(PipelineContext.kt:238)
	at io.ktor.util.pipeline.SuspendFunctionGun.loop(PipelineContext.kt:194)
	at io.ktor.util.pipeline.SuspendFunctionGun.access$loop(PipelineContext.kt:67)
	at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(PipelineContext.kt:144)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:272)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:79)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:54)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:36)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at WebSocketKt.Websockets(WebSocket.kt:7)
	at MainKt.main(main.kt:4)
```

**Comment by ToxicMushroom** [comment]
bump, had this issue today

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-340">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by 1354092549** [comment]
Call `client.webSocketSession` instead of `client.ws`, then you can control the session manually.
Note: you must close it to avoid resources leak.

**Comment by NurseyitTursunkulov** [comment]
@1354092549 send is not resolved. What to use istead?

**Comment by e5l** [comment]
Fixed in 1.5.1
